### DISCUSSION TITLE
Using zip files as containers for checkpoints, images and metadata

### DISCUSSION THREAD

**Discussion by Eljoseto** [vladmandic/sdnext#3314] [Ideas]
Hello, I have been wondering if it's possible to make an archive that includes the safetensor, the thumbnail image and the metadata txt or json to better organize the chekpoints folder.

The challenge would be to read the zip and load the checkpoint to memory.
Or changing the thumbnail image from SD.next and modifying the archive.

maybe there is a simpler solution, but anyway just want to know if tis possible.

**Comment by vladmandic** [comment]
to achieve best performance, model loading is not a sequential read process - it uses memory-mapped-file to load what it needs, when it needs - which means it needs direct access to safetensors file.
so any kind of archive would require that archive is unpacked before model can be loaded - and that is just huge i/o overhead.

what can be done is to store metadata and preview inside model's safetensors file itself, but i'm reluctant to modify model files that users download - that kind of behavior is never considered positive.

**Reply by Eljoseto** [reply]
Thanks for the quick reply, I guessed something of the sorts, but didn't know you coud store images inside a safetensors, could you please show me how to do this on my own?

**Reply by vladmandic** [reply]
take a look at safetensors save_file function: https://huggingface.co/docs/safetensors/api/torch#safetensors.torch.save_file  

note: you cannot store image as-is as only valid data types are either tensor or string. but you can encode image to base64 string and then store it.
